tools/webdriver/webdriver/bidi/client.py

```diff
@@ -0,0 +1,264 @@
+
+        for msg in self.send_buf:
+            await self._send(msg)
+
+    async def send(self, data):
+        if self.connection:
+            await self._send(data)
+        else:
+            self.send_buf.append(data)
+
+    async def _send(self, data):
+        msg = json.dumps(data)
+        print(msg)
+        await self.connection.send(msg)
+
+    async def handle(self, msg):
+        data = json.loads(msg)
+        await self.msg_handler(data)
+
+    async def end(self):
+        if self.connection:
+            await self.connection.close()
+            self.connection = None
+
+    async def read_messages(self):
+        async for msg in self.connection:
+            print(msg)
+            await self.handle(msg)
+
+
+class command:
+    """Decorator for implementing bidi commands
+
+    Implementing a command involves specifying an async function that
+    builds the parameters to the command. The decorator arranges those
+    parameters to be turned into a send_command call, using the class
+    and method names to determine the method in the call.
+
+    Commands decorated in this way don't return a future, but await
+    the actual response. In some cases it can be useful to
+    post-process this response before returning it to the client. This
+    can be done by specifying a second decorated method like
+    @command_name.result. That method will then be called once the
+    result of the original command is known, and the return value of
+    the method used as the response of the command.
+
+    So for an example, if we had a command test.testMethod, which
+    returned a result which we want to convert to a TestResult type,
+    the implementation might look like:
+
+    class Test(BidiModule):
+        @command
+        def test_method(self, test_data=None):
+            return {"testData": test_data}
+
+       @test_method.result
+       def _test_method(self, result):
+           return TestData(**result)
+    """
+
+    def __init__(self, fn):
+        self.params_fn = fn
+        self.result_fn = None
+
+    def result(self, fn):
+        self.result_fn = fn
+
+    def __set_name__(self, owner, name):
+        params_fn = self.params_fn
+        result_fn = self.result_fn
+
+        @functools.wraps(params_fn)
+        async def inner(self, **kwargs):
+            params = params_fn(self, **kwargs)
+            mod_name = owner.__name__.lower()
+            if hasattr(owner, "prefix"):
+                mod_name = "%s:%s" % (owner.prefix, mod_name)
+            cmd_name = "%s.%s" % (mod_name,
+                                  to_camelcase(name))
+            future = await self.session.send_command(cmd_name, params)
+            result = await future
+            if result_fn is not None:
+                result = result_fn(self, result)
+            return result
+
+        setattr(owner, name, inner)
+
+
+def to_camelcase(name):
+    parts = name.split("_")
+    parts[0] = parts[0].lower()
```

Comment:
Do you need to handle case with two consecutive underscores?

Reply:
I don't think so; it's only used to convert the method name into a command name.